File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans;
    public int sumRootToLeaf(TreeNode root) {
        if(root == null) return 0;
        
        dfs(root,0);
        return ans;
    }
    void dfs(TreeNode root,int val){
        if(root == null) return ;
        
        val = (val << 1) | root.val;
        
        if(root.left == null && root.right == null){
            ans += val;
        }
        dfs(root.left,val);
        dfs(root.right,val);
    }
}

File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    void dfs(TreeNode* root,int sum){
        if(!root) return ;
        
        sum  = (sum << 1) | root->val;
        
        if(!root->left && !root->right) ans += sum;
        
        dfs(root->left,sum);
        dfs(root->right,sum);
    }
    int sumRootToLeaf(TreeNode* root) {
        if(!root) return 0;
        if(!root->left && !root->right) return root->val;
        
        
        int sum = 0;
        
        dfs(root,sum);
        
        return ans;
    }
};